File: golang-devildyw/simple/L101对称二叉树.go
package main

func main() {

}

func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}
	// 还是老样子 将大的问题拆解为小的问题 轴对称图形 就需要轴两边对称位置的值是否一致了 自顶向下去比较
	// 现有函数结构肯定不太适用了 需要借助一个辅助函数

	return isSymmetricHelper(root, root)

}

func isSymmetricHelper(left *TreeNode, right *TreeNode) bool {
	if left == nil && right == nil {
		return true
	}

	if left == nil || right == nil {
		return false
	}

	return left.Val == right.Val && isSymmetricHelper(left.Left, right.Right) && isSymmetricHelper(left.Right, right.Left)

}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L101对称二叉树.java
package top.dyw.leetcode.Hot100;

public class L101对称二叉树 {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        }
        return isSymmetricHelper(root.left, root.right);
    }

    public boolean isSymmetricHelper(TreeNode left, TreeNode right) {
        if (left==null && right==null) {
            return true;
        }

        if (left==null || right==null) {
            return false;
        }

        return left.val == right.val && isSymmetricHelper(left.left, right.right) && isSymmetricHelper(left.right, right.left);
    }
}
